lib/client.js

```diff
@@ -2675,9 +2683,19 @@ MatrixClient.prototype.paginateEventTimeline = function(eventTimeline, opts) {
  * @return {module:http-api.MatrixError} Rejects: with an error response.
  */
 MatrixClient.prototype.login = function(loginType, data, callback) {
-    data.type = loginType;
+    var login_data = {
+        type: loginType,
+        device_id: this.deviceId,
+        initial_device_display_name: this.defaultDeviceDisplayName,
+    };
+
+    // merge data into login_data
+    for (var k in data) {
```

Comment:
I believe you can now use `for...of` to avoid having to do the `hasownProperty` mess